### PR TITLE
Improvements on console UI for install command

### PR DESCRIPTION
**Pull request by promatik**
### BEFORE - What was wrong? What was happening before this PR?

This PR aims part of the idea reported on https://github.com/Laravel-Backpack/ideas/issues/163.

### AFTER - What is happening after this PR?

Backpack installation is now much more complete, and in sync with the latest console UI changes on Laravel 9.21 — https://github.com/laravel/framework/pull/43065.

image

image

Click below to expand more previews;
<details>
  <summary>More previews</summary>
  image
  
  image
  
  image
  
  image
  
  image
</details>

https://user-images.githubusercontent.com/1838187/182003028-a2b648f5-b12d-47d5-9618-6730a86c77b2.mp4

### How did you achieve that, in technical terms?

Implementing manually the features we need from termwind (since it supports PHP8 only), it was not that hard since we need only a couple of them.

### Is it a breaking change?

No 👌

### How can we test the before & after?

Following the video above, and changing the branch of crud after it was installed.

### CODE REVIEW

**Comment by promatik** [comment]
Key points of this PR.

image

To add new addons to the the list, it's as easy as add the require command class to this list;

image

It must follow a pattern, but it's also a matter of copy past from the others.

**Comment by promatik** [comment]
Another important note, since this inclues installation of paid addons, every paid addon must make sure about the same stuff;
- composer repositories
- authentication

Since that was done in DevTools require command, I used the same code, which made everything MUCH simpler.

Each Require addon command class is now less than 100 lines, and is readable 😌👌

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [inline src/app/Console/Commands/Install.php:102]:
```suggestion
        $this->note('Currently there '.trans_choice("{0} are <fg=blue>no users</>|{1} is <fg=blue>1 user</>|[2,*] are <fg=blue>$currentUsers users</>", $currentUsers).' in the database.');
```

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [inline src/app/Console/Commands/Install.php:100]:
```suggestion
        $this->infoBlock('Create a user');
```

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [inline src/app/Console/Commands/Install.php:101]:
```suggestion
        $this->note('Quickly jump in your admin panel, using the email & password you choose here.');
```

File: src/app/Console/Commands/Addons/RequirePro.php
Comment (tabacitu) [inline src/app/Console/Commands/Addons/RequirePro.php:35]:
```suggestion
            'Adds 5 operations, 10 filters, 28 fields, 6 columns, charts',
```

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [inline src/app/Console/Commands/Install.php:86]:
```suggestion
        $this->note("Head to <fg=blue>$url</> to access your new admin panel.");
```

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [inline src/app/Console/Commands/Install.php:87]:
```suggestion
        $this->note('You may need to run `php artisan serve` to serve your Laravel project.');
```

File: src/BackpackServiceProvider.php
Comment (tabacitu) [inline src/BackpackServiceProvider.php:23]: My biggest problem with this PR - even though everything works fine, I keep getting this weird error upon each command execution. Previously (on `main`) the output was nice and clean:
CleanShot 2022-07-31 at 08 45 11@2x

But in this PR, I get `sh: mode: command not found` all the freakin' time:

CleanShot 2022-07-31 at 08 48 25@2x

Any idea how to solve this @promatik ?

**Review by tabacitu — CHANGES_REQUESTED** [review]
I FREAKING LOVE THIS! ❤️❤️❤️ Well done @promatik ! 

Beautiful & functional 👏👏👏

I want to merge this ASAP, so let's clean it up please 🙏 I have one big problem with it (the error I'm getting), other than that, I just changes the texts here and there. I want to test again after that error is fixed.

File: src/BackpackServiceProvider.php
Comment (promatik) [reply src/BackpackServiceProvider.php:23]: This should now be fixed 👌
I took a look at termwind code 🤫 to see how they get the the console size 👌

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [inline src/app/Console/Commands/Install.php:100]:
```suggestion
        $this->infoBlock('Creating an admin:');
```

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [inline src/app/Console/Commands/Install.php:167]:
```suggestion
        $this->infoBlock('Installing premium Backpack add-ons:');
```

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [inline src/app/Console/Commands/Install.php:168]:
```suggestion
        $this->note('Add tons of features and functionality to your admin panel, using our paid add-ons.');
```

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [inline src/app/Console/Commands/Install.php:169]:
```suggestion
        $this->note('For more information, payment and access please visit https://backpackforlaravel.com/pricing');
```

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [inline src/app/Console/Commands/Install.php:183]:
```suggestion
            $input = (int) $this->listChoice('Would you like to install a premium Backpack add-on? <fg=gray>(enter option number)</>', $this->addons->toArray());
```

File: src/app/Console/Commands/Addons/RequirePro.php
Comment (tabacitu) [inline src/app/Console/Commands/Addons/RequirePro.php:49]: Hmmm... I don't know if this info block is a good idea here:
- 20% use case - if you run this command individually - maybe it makes sense; though... you know what's going to happen, you typed the command, so even then it might be superfluous;
- 80% use case - if you run this command as part of `php artisan backpack:install`, the info block makes this command look like it's at the same level as `Install Backpack add-ons?` which is a bit confusing:

CleanShot 2022-08-01 at 08 09 52

The way I see it as a user, installing one of the premium add-ons is part of the question "_Would you like to install a Backpack add-on?_" so nothing I do there should be at the same level as that question, but _part of it_. Right?

Comment (tabacitu) [reply src/app/Console/Commands/Addons/RequirePro.php:49]: Additionally, I love the line that says "running" while installing it, but after it's done, the "Running" line transforms to "DONE", which is again superfluous because there will be a list below with all premium add-ons, where the current add-on will be marked as INSTALLED.

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [inline src/app/Console/Commands/Install.php:85]: I see you've changed the blue INFO label to DONE here - didn't know we can do that. In this case, why don't we move all `infoBlock()` calls to the `php artisan backpack:install` command, and name them `Step 1`, `Step 2`, `Step 3`? That way:
- if you run that command individually, you won't have info blocks (you don't really need them, you know what that command is doing already, it's in the name);
- if you run that command as part of `php artisan backpack:install`, it'll read as a smooth, step-by-step process:
CleanShot 2022-08-01 at 08 21 45

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [inline src/app/Console/Commands/Install.php:86]:
```suggestion
        $this->note("Go to <fg=blue>$url</> to access your new admin panel.");
```

**Review by tabacitu — CHANGES_REQUESTED** [review]
I LOVE IT! I love it, I love it, I love it! ❤️❤️❤️ Thanks @promatik !

I'm noticing a few design/visual changes we should be making here, to give the whole install process a coherent look & feel, mainly in terms of visual hierarchy.

But other than that... this is working WONDERFUL for me 👏👏👏
Much MUCH better than before.

File: src/app/Console/Commands/Addons/RequirePro.php
Comment (tabacitu) [reply src/app/Console/Commands/Addons/RequirePro.php:49]: See my next comment and image, please, that will be even more explicit.

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [inline src/app/Console/Commands/Install.php:164]: Haha I just noticed this when running `php artisan backpack:install` on a project with everything already installed. Nice touch! 👏👏👏

**Comment by tabacitu** [comment]
I FREAKING LOVE THIS! EXCELLENT WORK @promatik !!! 

happy_dance_cap

@pxpm could you please double-check? I want to launch this ASAP.

File: src/app/Console/Commands/Traits/AddonsHelper.php
Comment (pxpm) [inline src/app/Console/Commands/Traits/AddonsHelper.php:103]: We know we have some "pitfalls" when user adds wrong authentication, and then it's cached and then you need to dabble a little bit in your brain how to fix it (the issues we found in the other PR this one replaced).

I think it would atleast help if we do minimal validation here like, `not empty` and atleast `min:xx` characters (the number configured in backpack application to generate tokens).

It will not solve all the problems but atleast will prevent  "simple" mistakes:
- pressing enter twice;
- not copying correctly from backpack website;
- not pasting correctly on the console;

File: src/app/Console/Commands/Install.php
Comment (pxpm) [inline src/app/Console/Commands/Install.php:89]: using the backticks here does not make `php artisan serve` standout. We should probably add some color/bold whatever to it.

File: src/app/Console/Commands/Install.php
Comment (pxpm) [inline src/app/Console/Commands/Install.php:99]: Please let's take into consideration that users could contain millions of records and this `count()` would be a very expensive operation in that case.

I was about to suggest to use an "optimized query count" here, but thinking better about it, **we should remove it completely**. 😿 
I know it's a very nice feedback to give user the the amout of users in database, but I don't think it's `necessary/mandatory` thing here, it's just "nice", and we are exposing ourselfs to future problems. 

Just letting the user know that their desired user is created and they can login should be enough.

File: src/app/Console/Commands/Install.php
Comment (pxpm) [inline src/app/Console/Commands/Install.php:116]: Let's take into consideration that user migth be using a different hashing algorithm, or have configured their algorithm. 

We should probably be using `Hash::make` here. https://laravel.com/docs/9.x/hashing

File: src/app/Console/Commands/Traits/AddonsHelper.php
Comment (promatik) [reply src/app/Console/Commands/Traits/AddonsHelper.php:103]: Nice idea, will do it 👌
I'll add simple validations to all the inputs.

File: src/app/Console/Commands/Install.php
Comment (pxpm) [inline src/app/Console/Commands/Install.php:117]: We could check if model uses timestamps before assuming it does. 

`if($userModel->timestamps) { .. }`

File: src/app/Console/Commands/Install.php
Comment (promatik) [reply src/app/Console/Commands/Install.php]: I agree 👌
image

That one was easy 😅

File: src/app/Console/Commands/Install.php
Comment (promatik) [reply src/app/Console/Commands/Install.php:117]: Nice catch 👌

File: src/app/Console/Commands/Install.php
Comment (promatik) [reply src/app/Console/Commands/Install.php:116]: Another nice catch 😅

File: src/app/Console/Commands/Install.php
Comment (promatik) [reply src/app/Console/Commands/Install.php:99]: It will be SUPER rare the developer has millions of entries here 😅

I think this is useful in cases where, developer is trying out Backpack, and in the between of the installation process, he is playing around and created an user ...

---

Anyway, @pxpm is there any possible fix for it?
Can we move it inside a try catch block?

File: src/app/Console/Commands/Install.php
Comment (pxpm) [inline src/app/Console/Commands/Install.php:176]: Do your magic trick with the colors here in this link too @promatik 🌈

File: src/app/Console/Commands/Install.php
Comment (promatik) [reply src/app/Console/Commands/Install.php]: Done 👌

File: src/app/Console/Commands/Install.php
Comment (pxpm) [reply src/app/Console/Commands/Install.php:99]: I don't think it matter at all in this case if he created users or not in the meanwhile. 

He **wanted to create an admin user**, he got the admin user created and can now login with it. 

I agree that is not everyday that you will be having millions of users, unless in that day you are migrating a legacy app to backpack and already have the users setup 👍 

There is no benefit that I can see here by having the user `count()` sorry 😿

File: src/app/Console/Commands/Install.php
Comment (promatik) [reply src/app/Console/Commands/Install.php]: Also done 👌

File: src/app/Console/Commands/Traits/AddonsHelper.php
Comment (tabacitu) [reply src/app/Console/Commands/Traits/AddonsHelper.php:103]: I agree. But let's make a separate issue please. This doesn't relate to the current PR, it's an unrelated improvement.

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [reply src/app/Console/Commands/Install.php]: I like the `count()`, let's keep it. Like @promatik said, I think it's unlikely for the dev to have millions of users on localhost.

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [reply src/app/Console/Commands/Install.php:182]: Yes please!

**Comment by tabacitu** [comment]
In regards to Pedro's comment above:

> 1) I think it would be usefull to have default values for the user questions (name/email/password).

I'd rather not add defaults at all. I don't want people creating the stock admin, and knowing you can "probably" try that email or email/password combo in all Backpack admin panels. Let them choose their own admin & password, no defaults.

> 2) I think it would help in the "resume of created user" to include the password too so it could be copy-pasted:

Maaaybe... I don't know... I'll let @promatik decide about this one. Feels odd to echo the password, when the field itself will not echo the password.

> 3) Would it be cumbersome to add some validation ? Nothing fancy, just like:

Yeah we should probably add validation. Bonus points if we re-use the validation from Registration or something. Then again this would make the whole process a lot more difficult, maybe you actually WANT to add an admin on localhost that is "admin" and "password" just so it's easy. It's localhost after all. So maybe we just add `required` as validation? Not a prio, but let's do it in a separate issue please.

> 4) Wouldn't be better to add some more "focus" on the numbers for the addons ? Maybe indent them inside the question and add some [1] or (1) or « 1 » -

Yeah I also wish the numbers would be a little more evident.

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [inline src/app/Console/Commands/Install.php]: @promatik - heads-up, I like the old message better, to be honest. And let's keep the `count()` please, I liked it. Maybe the end result looks something like this?

> Currently there are _2 users_ in the database.

The expanded version would be:

> Currently there are _2 users_ in the database (`users` table, as defined in `App\Models\User`).

But I think that's too much information for this point in time. Let's keep it simple here.

File: src/app/Console/Commands/Install.php
Comment (pxpm) [reply src/app/Console/Commands/Install.php]: Errr .. !! @tabacitu you like the "displayed interface", but please take into consideration the underlaying issue and the benefits it brings. 
When that dev with 400k users, an un-optimized db, tries to use backpack on his project, he will see big memory consumption and loading times, for what should be an "easy sql insert".

It can take 10s, no memory limit hit and we are all good, or it can hit timeout / memory limit and break the install script. 
My question is: for Why leave this door open praying for sunny days when it does not add any benefit for the dev experience? 

When user is created we **already give feedback**: User created with name x, email x, pass x ! Add another one ? 

My last arguments, I rest my case here.

**Comment by tabacitu** [comment]
@promatik let's wrap this up, I want to merge and tag it ASAP please 🙏 Please add separate issues for any other suggestions here, and we'll do them one-by-one. Nothing I see here is worth postponing the merge, in my opinion - not even my suggestions.

To rephrase the importance: Every day we delay merging this, we offer a sub-par installation experience to hundreds of devs. When we have a much much better one ready!

Is this ready for me to click the merge button?

obsessive-compulsive-animation-gif-by-michael-frei

**Comment by pxpm** [comment]
Apart from what I already wrote, this is working fine, you can hit that button ✅

File: src/app/Console/Commands/Install.php
Comment (promatik) [reply src/app/Console/Commands/Install.php]: Done 👌

File: src/app/Console/Commands/Install.php
Comment (promatik) [reply src/app/Console/Commands/Install.php]: Reverted 👌

**Comment by promatik** [comment]
@tabacitu @pxpm the use count message is reverted.
This is ready to merge, I think 🙌

I'll add a PR with WIP for everything discussed here;

- [ ] Validate user fields
- [ ] Validate auth.json token fields
- [ ] Generate a password in case user left it blank (only in that case, show it to the user)

File: src/app/Console/Commands/Install.php
Comment (tabacitu) [reply src/app/Console/Commands/Install.php:100]: I don't think counting 400k users will take 10s. I don't think it will even take 1s. And the likelyhood for the db to be unoptimized is slim - after all, we're expecting the standard Laravel `users` table.

I'm afraid I don't agree here Pedro - let's use `count()` for now, and if it turns out I'm wrong (which I often am!) someone will report it, and we can remove or optimize it. But for the 90% use case, I think this will work just fine.

**Comment by tabacitu** [comment]
Writing about this in our Progress Report. So so happy with it ❤️ Just look at this beauty:
CleanShot 2022-08-19 at 11 42 06

**Pull request merged**
